engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java

```
@@ -780,17 +780,17 @@ public void monitorHeartbeat(boolean waitForRecovery)
           //LOG.debug("{} {} {}", c.getExternalId(), currentTms - sca.createdMillis, this.vars.heartbeatTimeoutMillis);
           // container allocated but process was either not launched or is not able to phone home
           if (currentTms - sca.createdMillis > 2 * this.vars.heartbeatTimeoutMillis) {
-            LOG.error("Container {}@{} startup timeout ({} ms).", c.getExternalId(), c.host, currentTms - sca.createdMillis);
+            LOG.warn("Container {}@{} startup timeout ({} ms).", c.getExternalId(), c.host, currentTms - sca.createdMillis);
             containerStopRequests.put(c.getExternalId(), c.getExternalId());
           }
         } else {
           if (currentTms - sca.lastHeartbeatMillis > this.vars.heartbeatTimeoutMillis) {
             if (!isApplicationIdle()) {
               // Check if the heartbeat for this agent has already been missed to raise the StramEvent only once
               if (sca.lastHeartbeatMillis != -1) {
-                String info = String.format("Container %s@%s heartbeat timeout (%d%n ms).", c.getExternalId(), c.host, currentTms - sca.lastHeartbeatMillis);
-                LOG.error(info);
-                StramEvent stramEvent = new StramEvent.ContainerErrorEvent(c.getExternalId(), info, null);
+                String info = String.format("Container %s@%s heartbeat timeout  (%d%n ms).", c.getExternalId(), c.host, currentTms - sca.lastHeartbeatMillis);
+                LOG.warn(info);
+                StramEvent stramEvent = new StramEvent.ContainerErrorEvent(c.getExternalId(), info, null, StramEvent.LogLevel.WARN);
```

@Hitesh-Scorpio I'd suggest waiting for @tweise to respond before making code changes. The timeout here has quite significant impact on application and will lead to downstream operators being re-deployed.

---

The "stram" event is a user facing concept and logging is internal. For the user the container not starting or failing is probably means "error", for Apex engine it is not. Error logging should be appropriate to diagnose issues with Apex, and hence the log4j level cannot be error (it isn't an error in the engine).  If what we set on the event is called "log level", then perhaps it should be the same, but I don't see how that information is important to the user. I would suggest the event to have its own severity that should not be coupled to log4j logging.

---

logs are created not for the apex engine, but for dev-ops and apex developers to analyze and troubleshoot problems both in the engine itself and in the run-time environments. It will be confusing for users to get ContainerErrorEvent and not to have any indication of the problem in the AppMaster logs.

 To me, this is similar to a checked exception in Java. Checked exceptions are handled by an application, but they are still exceptions and need to be logged as errors or warnings.

I was not referring to the StramEvent level, the event that is raise is Container*Error*Event and it should use default level for the Event.

---

The event is not due to an internal error or condition that cannot be handled by the system and hence ERROR level in log4j isn't appropriate. This has been discussed before and we don't use logging at this level for similar conditions elsewhere (at least I hope that is still the case). 

Exception handling: If an exception is "handled", then there is normally no need to also log it (at least not as error). Such logging usually occurs for exceptions that are not handled in a catch all handler, where there is no other action that can be taken.
 
I don't want to go in circles. If we cannot reach agreement then the logger change needs to be reverted as part of the original JIRA.

---

@tweise Do you agree to change both LOG levels to WARN? I disagree with INFO level, but my original suggestion was to change it to WARN (or ERROR), so I am OK with WARN.

---

Yep, WARN looks reasonable.